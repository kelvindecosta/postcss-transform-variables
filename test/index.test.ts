import fs from 'fs';

import MurmurHash from 'imurmurhash';
import postcss from 'postcss';

import plugin from '../dist';
import type { Options } from '../dist';

const loadCSSFromCase = (name: string) =>
	fs.readFileSync(`${__dirname}/__fixtures__/${name.replace(/\s+/g, '-')}.css`, 'utf8');

describe('options', () => {
	const run = (options?: Options) => () =>
		postcss([plugin(options)]).process('', { from: undefined });

	it('can be omitted', () => {
		expect(run()).not.toThrow();
		expect(run({})).not.toThrow();
	});
});

describe.each<{ name: string; transform: Options['transform'] }>([
	{
		name: 'without a double hyphen',
		transform: ({ identifier }) => `kd-${identifier}`
	},
	{
		name: 'with a double hyphen',
		transform: ({ identifier }) => `--kd-${identifier}`
	}
])('transform $name', ({ transform }) => {
	const run = (input: string) =>
		postcss([plugin({ transform })]).process(input, { from: undefined }).css;

	const cases = [
		'a single variable',
		'a single variable used multiple times',
		'a single variable with a fallback',
		'a single variable using another',
		'multiple variables',
		'multiple variables in a single property',
		'multiple variables inside functions'
	];

	it.each(cases)('should handle %s', (type) => {
		const input = loadCSSFromCase(type);

		expect(run(input)).toMatchSnapshot();
	});
});

describe('collisions', () => {
	const input = loadCSSFromCase('multiple variables');
	const transform: Options['transform'] = ({ identifier }) => `kd-${identifier?.substring(0, 2)}`;

	const run = (warnOnDetectCollision: Options['warnOnDetectCollision']) => () =>
		postcss([plugin({ transform, warnOnDetectCollision })]).process(input, {
			from: undefined
		});

	it('can be logged as warnings', () => {
		const result = run(true)();
		expect(result.warnings().length).toBe(1);
		expect(result.css).toMatchSnapshot();
	});
	it('can be ignored', () => {
		const result = run(false)();
		expect(result.warnings().length).toBe(0);
		expect(result.css).toMatchSnapshot();
	});
});

describe('non-determinism', () => {
	const input = loadCSSFromCase('a single variable used multiple times');
	const hash = MurmurHash();
	const transform: Options['transform'] = ({ identifier }) =>
		`kd-${hash
			.hash(identifier as string)
			.result()
			.toString(16)}`;

	const run = (warnOnDetectNonDeterminism: Options['warnOnDetectNonDeterminism']) => () =>
		postcss([plugin({ transform, warnOnDetectNonDeterminism })]).process(input, {
			from: undefined
		});

	it('can be logged as warnings', () => {
		const result = run(true)();
		expect(result.warnings().length).toBe(1);
		expect(result.css).toMatchSnapshot();
	});
	it('can be ignored', () => {
		const result = run(false)();
		expect(result.warnings().length).toBe(0);
		expect(result.css).toMatchSnapshot();
	});
});
