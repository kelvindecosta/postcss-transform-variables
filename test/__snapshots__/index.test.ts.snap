// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collisions can be ignored 1`] = `
":root {
  --kd-re: red;
  --kd-bl: blue;
  --kd-bl: black;
}

p {
  color: var(--kd-re);
  background-color: var(--kd-bl);
}

a {
  color: var(--kd-bl);
}
"
`;

exports[`collisions can be logged as warnings 1`] = `
":root {
  --kd-re: red;
  --kd-bl: blue;
  --black: black;
}

p {
  color: var(--kd-re);
  background-color: var(--black);
}

a {
  color: var(--kd-bl);
}
"
`;

exports[`non-determinism can be ignored 1`] = `
":root {
  --kd-ef934a92: red;
}

p {
  color: var(--kd-c7d1e12d);
}

blockquote {
  color: var(--kd-560c096c);
}
"
`;

exports[`non-determinism can be logged as warnings 1`] = `
":root {
  --kd-eb5026fd: red;
}

p {
  color: var(--kd-eb5026fd);
}

blockquote {
  color: var(--kd-eb5026fd);
}
"
`;

exports[`transform with a double hyphen should handle a single variable 1`] = `
":root {
  --kd-red: red;
}

p {
  color: var(--kd-red);
}
"
`;

exports[`transform with a double hyphen should handle a single variable used multiple times 1`] = `
":root {
  --kd-red: red;
}

p {
  color: var(--kd-red);
}

blockquote {
  color: var(--kd-red);
}
"
`;

exports[`transform with a double hyphen should handle a single variable using another 1`] = `
":root {
  --kd-red: red;
}

a {
  --kd-link-color: var(--kd-red);
  color: var(--kd-link-color);
}
"
`;

exports[`transform with a double hyphen should handle a single variable with a fallback 1`] = `
":root {
  --kd-red: red;
}

p {
  color: var(--kd-red, pink);
}
"
`;

exports[`transform with a double hyphen should handle multiple variables 1`] = `
":root {
  --kd-red: red;
  --kd-blue: blue;
  --kd-black: black;
}

p {
  color: var(--kd-red);
  background-color: var(--kd-black);
}

a {
  color: var(--kd-blue);
}
"
`;

exports[`transform with a double hyphen should handle multiple variables in a single property 1`] = `
":root {
  --kd-padding-y: 2px;
  --kd-padding-x: 4px;
}

p {
  padding-block: var(--kd-padding-y) var(--kd-padding-x);
}
"
`;

exports[`transform with a double hyphen should handle multiple variables inside functions 1`] = `
":root {
  --kd-h: 200;
  --kd-s: 50%;
  --kd-l: 50%;
}

a {
  color: hsl(var(--kd-h) var(--kd-s) var(--kd-l));
}
"
`;

exports[`transform without a double hyphen should handle a single variable 1`] = `
":root {
  --kd-red: red;
}

p {
  color: var(--kd-red);
}
"
`;

exports[`transform without a double hyphen should handle a single variable used multiple times 1`] = `
":root {
  --kd-red: red;
}

p {
  color: var(--kd-red);
}

blockquote {
  color: var(--kd-red);
}
"
`;

exports[`transform without a double hyphen should handle a single variable using another 1`] = `
":root {
  --kd-red: red;
}

a {
  --kd-link-color: var(--kd-red);
  color: var(--kd-link-color);
}
"
`;

exports[`transform without a double hyphen should handle a single variable with a fallback 1`] = `
":root {
  --kd-red: red;
}

p {
  color: var(--kd-red, pink);
}
"
`;

exports[`transform without a double hyphen should handle multiple variables 1`] = `
":root {
  --kd-red: red;
  --kd-blue: blue;
  --kd-black: black;
}

p {
  color: var(--kd-red);
  background-color: var(--kd-black);
}

a {
  color: var(--kd-blue);
}
"
`;

exports[`transform without a double hyphen should handle multiple variables in a single property 1`] = `
":root {
  --kd-padding-y: 2px;
  --kd-padding-x: 4px;
}

p {
  padding-block: var(--kd-padding-y) var(--kd-padding-x);
}
"
`;

exports[`transform without a double hyphen should handle multiple variables inside functions 1`] = `
":root {
  --kd-h: 200;
  --kd-s: 50%;
  --kd-l: 50%;
}

a {
  color: hsl(var(--kd-h) var(--kd-s) var(--kd-l));
}
"
`;
